### PR TITLE
Advanced Building Blocks - Bubble Sort

### PR DESCRIPTION
**Pull request by mricanho**
We created bubble_sort and bubble_sort_by methods.

We used the bubble sort algorithm and for that, we applied booleans and loops

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#     swap = false
+#     number_of_loops.times do |x|
+#       arr[x], arr[x + 1] = arr[x + 1], arr[x] if arr[x] > arr[x + 1]
+#       swap = true
+#     end
+#     number_of_loops -= 1
+#   end
+#   arr
+# end
+# puts bubble_sort([3, 2, 1, 85, 1, 58, 6, 4])
```
Comment (SpaYco) [inline main.rb]: - [x] remove commented code ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/))

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      arr[j], arr[j + 1] = arr[j + 1], arr[j] if arr[j].length > arr[j + 1].length
+    end
+  end
+  arr
+end
+
+compare = bubble_sort_by(%w[hi hello hey]) do |left, right|
+  left.length - right.length
+end
+puts compare
```
Comment (SpaYco) [inline main.rb]: - [x] No debugging instructions in committed code (as discussed [here](https://questions.microverse.org/t/debugging-instructions-code/346)) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/section/debugging_requirement.png) from Pathwright)

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+def bubble_sort_by(arr)
+  number_ofloop = arr.length - 1 # loop inside the index not the length
+  (0..number_ofloop).each do |_i|
+    (0...number_ofloop).each do |j|
+      arr[j], arr[j + 1] = arr[j + 1], arr[j] if arr[j].length > arr[j + 1].length
+    end
+  end
+  arr
+end
```
Comment (SpaYco) [inline main.rb]: - [x] #bubble_sort_by isn't working, if you try `bubble_sort_by(%w[hi hello hey]) { |left, right| left.length + right.length }` it wouldn't give the right result (`[hello hey hi]`) (Hint: you're not using `yield`)

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

## Required Changes ♻️ 

- [x] check below 👇 

### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+def bubble_sort(arr)
+  number_ofloop = arr.length - 1
+  (0..number_ofloop).each do |_i|
```
Comment (kitpao) [inline main.rb]: - [x] **OPTIONAL** It is a convention in Ruby to use the only `_` for unused parameters, or you could skip the parameters as well and leave it like: `(0..number_ofloop).each do`

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  arr
+end
+
+puts bubble_sort([4, 3, 78, 2, 0, 2])
+
+# The second:
+
+def bubble_sort_by(arr)
+  number_ofloop = arr.length - 1 # loop inside the index not the length
+  (0..number_ofloop).each do |_i|
```
Comment (kitpao) [inline main.rb]: - [x] **OPTIONAL** It is a convention in Ruby to use the only `_` for unused parameters, or you could skip the parameters as well and leave it like: `(0..number_ofloop).each do`

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  number_ofloop = arr.length - 1
+  (0..number_ofloop).each do |_i|
+    (0...number_ofloop).each do |j|
+      arr[j], arr[j + 1] = arr[j + 1], arr[j] if arr[j] > arr[j + 1]
+    end
+  end
+  arr
+end
+
+puts bubble_sort([4, 3, 78, 2, 0, 2])
```
Comment (kitpao) [inline main.rb]: - [x] Please delete this line of debugging code (in the future you will learn to make automated tests for these cases). For now, if you want, you could save your examples in a different file and gitignore it (you will find it useful for enumerables project)
image

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @mricanho and @AhmedAmin90 

Good job making your methods pass all the tests we have. You followed the principle of bubble sort correctly. 👏 You have one technical detail, please check below 🔽 

**_Note:_** All the **[OPTIONAL]** are **not required** in order to approve your project but are personal suggestions to improve your coding experience

If you are curious about optimizing your code, you can take a look at the resulting code at the end of [this article](https://medium.com/swlh/bubble-sort-algorithm-how-to-develop-your-computational-thinking-c092a9b4bf16)

### ♻️ Status: Your project requires 1 additional change ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Review by kitpao — APPROVED** [review]
# 👋 Hello @AhmedAmin90 and @mricanho 

Great job 💯 your project complies with all requirements 🥳 

### 👏 Your project has been approved 👏
### 🚀 You may merge to main and submit the project completition form 🚀

___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Pull request merged**